From 433b883c007736bed5c4df0b9db4d3ff498f9288 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Mon, 23 Feb 2026 14:50:27 +0800
Subject: [PATCH] d2d1: Clamp opacity value for brushes.

Some applications call DrawBitmap() with the opacity value set to 255.0f and end up showing white in
the end result. Tests show that the opacity value should be clamped to legitimate values. Command
lists store the original opacity value according to tests.

(cherry picked from commit 8990db8f714f6b76c0f0de3529deb5d317d1121f)
---
 dlls/d2d1/brush.c      | 2 +-
 dlls/d2d1/tests/d2d1.c | 4 ----
 2 files changed, 1 insertion(+), 5 deletions(-)

diff --git a/dlls/d2d1/brush.c b/dlls/d2d1/brush.c
index 50754a2f21a..ea09e6eebbd 100644
--- a/dlls/d2d1/brush.c
+++ b/dlls/d2d1/brush.c
@@ -1385,7 +1385,7 @@ BOOL d2d_brush_fill_cb(const struct d2d_brush *brush, struct d2d_brush_cb *cb)
     }
 
     cb->type = brush->type;
-    cb->opacity = brush->opacity;
+    cb->opacity = max(0.0f, min(brush->opacity, 1.0f));
 
     switch (brush->type)
     {
diff --git a/dlls/d2d1/tests/d2d1.c b/dlls/d2d1/tests/d2d1.c
index c0e778c4b66..35de21e6e13 100644
--- a/dlls/d2d1/tests/d2d1.c
+++ b/dlls/d2d1/tests/d2d1.c
@@ -2546,7 +2546,6 @@ static void test_color_brush(BOOL d3d11)
 
     get_surface_readback(&ctx, &rb);
     colour = get_readback_colour(&rb, 0, 0);
-    todo_wine
     ok(compare_colour(colour, 0xff7f0000, 1), "Got unexpected colour 0x%08lx.\n", colour);
     colour = get_readback_colour(&rb, 1, 0);
     ok(compare_colour(colour, 0xff010000, 1), "Got unexpected colour 0x%08lx.\n", colour);
@@ -2972,7 +2971,6 @@ static void test_bitmap_brush(BOOL d3d11)
 
     get_surface_readback(&ctx, &rb);
     colour = get_readback_colour(&rb, 0, 0);
-    todo_wine
     ok(compare_colour(colour, 0xff7f0000, 1), "Got unexpected colour 0x%08lx.\n", colour);
     colour = get_readback_colour(&rb, 1, 0);
     ok(compare_colour(colour, 0xff010000, 1), "Got unexpected colour 0x%08lx.\n", colour);
@@ -2993,7 +2991,6 @@ static void test_bitmap_brush(BOOL d3d11)
 
     get_surface_readback(&ctx, &rb);
     colour = get_readback_colour(&rb, 0, 0);
-    todo_wine
     ok(compare_colour(colour, 0xff7f0000, 1), "Got unexpected colour 0x%08lx.\n", colour);
     colour = get_readback_colour(&rb, 1, 0);
     ok(compare_colour(colour, 0xff010000, 1), "Got unexpected colour 0x%08lx.\n", colour);
@@ -3023,7 +3020,6 @@ static void test_bitmap_brush(BOOL d3d11)
 
     get_surface_readback(&ctx, &rb);
     colour = get_readback_colour(&rb, 0, 0);
-    todo_wine
     ok(compare_colour(colour, 0xff7f0000, 1), "Got unexpected colour 0x%08lx.\n", colour);
     colour = get_readback_colour(&rb, 1, 0);
     ok(compare_colour(colour, 0xff010000, 1), "Got unexpected colour 0x%08lx.\n", colour);
